src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -754,7 +750,13 @@ private void GenerateFastPathFuncForObject(SourceWriter writer, ContextGeneratio
                     _propertyNames.TryAdd(runtimePropName, propNameVarName);
 
                     DefaultCheckType defaultCheckType = GetDefaultCheckType(contextSpec, propertyGenSpec);
-                    string? objectExpr = castingRequiredForProps ? $"(({propertyGenSpec.DeclaringType.FullyQualifiedName}){ValueVarName})" : ValueVarName;
+
+                    // For properties whose declared type differs from that of the serialized type
+                    // perform an explicit cast -- this is to account for hidden properties or diamond ambiguity.
+                    string? objectExpr = propertyGenSpec.DeclaringType != typeGenSpec.TypeRef
+                        ? $"(({propertyGenSpec.DeclaringType.FullyQualifiedName}){ValueVarName})"
```

Comment:
>? $"(({propertyGenSpec.DeclaringType.FullyQualifiedName}){ValueVarName})" [](http://example.com/codeflow?start=24&length=73)

I assume the cast in this case will never fail at runtime. right?

Reply:
It shouldn't, the serialized value is always a subtype of the declaring type of the property.